### ISSUE TITLE
Request for features :)

### ISSUE THREAD

**Issue by alexjp** [psifidotos/nowdock-plasmoid#22]
Hi,

Thanks for the very nice nowdock. ( and for workflow project a few years ago )

Its almost what I need. I am trying to do this:

http://imgur.com/KHOf1KE
and 
http://imgur.com/ZEzqocp

meaning, one dock per activity, with different launchers.

Would be great if could request the following features:
- Anchor dock at the bottom middle always
- Save default settings ( so that is easy to add docks to new activities )
- Make only show pinned apps
- Allow to rearrange position of apps on dock by dragging

Could these be possible ? 

Thanks !!!!

**Comment by psifidotos** [comment]
Thank Alex!!! :) 

your kind words give me strength for the future...
For your requests, I am a little confused, I think all of them are supported, do I miss something?:
1. You can add the plasmoid in your desktop, meaning that each activity can have different launchers set and can show only per-activity windows. So each activity can show its specific launchers+only its windows). Issue here is that if you put the plasmoid in the panel to be able to be on top of windows this can not be done I think, but now that I am thinking it :), you put flies in my head :). Give some more info around this what you do want... :)
2. Of course, this is already there...
3. Needs info, can you explain me please?
4. Of course, this is already there. But you must lock your desktop first in order to be able to drag tasks around. Both dragging in supported and also to drag n' drop from outside an app in order to add a new launcher...

I hoped this has helped...

**Comment by alexjp** [comment]
Hi !
So ... explanations below :)
1. Yes, currently I am testing it like that and it works... that was the part of me explaining my goal, was not actually a feature request.
2. Save default settings, I don't think its there. What i mean is if i configure a dock and press save as default button, next created nowDock will have its settings ( 64px size, specific manifiying size, specific behaviour settings ). I if want all nowDocks to have 64px size, and 1.3 magnifiying and specific panel transparency and size, its a major bummer to have to configure everyone of them on each activity to look alike.
3. Currently nowDock shows pinned apps and other apps. So if i have a pinned app on nowDock : dolphin, and open konsole, it will show dolphin icon and konsole icon. Pinned apps only, it would only show dolphin icon and hide konsole icon.
4. It is already there, just tried it !!! I always had desktop unlock when trying to rearrange them ! sorry !

Thanks :)

**Comment by psifidotos** [comment]
So Alex,

happy to hear 1,4 is working!!!!  :)

concerning 2,3...
2. Change - Save default settings. I believe this is not an issue with Now Dock, defaults can not be changed by the user in any application or widget. So this is an upstream issue-question (for the plasma developers approach in plasma 5 to solve this).
In Plasma 4 era there were two ways to achieve this and not reconfigure everything from start.
2.1 One way was to copy an activity, the new activity was exactly like the first one with all the widgets keeping their settings. But I think that this functionality has been dropped by the plasma developers.
2.2 The other way was with Activities Templates. It was possible to create a new activity based on a template which was setting all the needed values and positions for its widgets. But I dont see such a choice for the user in Plasma 5.

3.There is an issue in such behavior. In your example, the konsole icon is not shown and the konsole window it does. If you minimize your konsole window then the only way to access it, is through Alt->Tab or through Present Windows Kwin effect and there is also an issue that there is not a visual to indicate that such a window exists.
Could you please explain me what are you trying to achieve? For example, why you dont want to see the konsole window even though is shown in your activity?

cheers,
michail

**Comment by psifidotos** [comment]
Alex, update for 2...
why dont you add the Now Dock plasmoid in a bottom panel in order to use it in all your activities?

**Comment by alexjp** [comment]
Hi,

Yes, point 2, the previous kde4/plasma1 functionality would do it. Greatly miss it.
1. What i was trying to achieve with this would be to have the dock as the "important activity apps only" and then have the normal taskbar on a hidden panel where it would show all other windows of the activity.

On update for 2... I don't want to use the bottom panel dock, because i don't want the same dock on every activity. The real major point is to have a specific apps and launchers dock for each activity. There is even an activity where i don't have any launchers/dock.

Having specific launchers for each activity "on the same dock plasmoid" would fix this, but its currently impossible.

Thank you. If you want to talk more in detail, just let me know, we can talk by IRC or some other way. Let me know if i can help in any way

**Comment by psifidotos** [comment]
> Having specific launchers for each activity "on the same dock plasmoid" would fix this, but its currently impossible.

Well, this is impossible until tomorrow or late at night today... :-)
I am going to support this, because I also needed...
For each launcher there will be a choice, "Launcher On All Activities" (enabled by default when you add a launcher), but by unchecking it this launcher will be available only on the current activity, or any other activity you will add it...

> Let me know if i can help in any way

Help is always welcome!! :)
I dont know what it is currently most needed... in what areas can you help????? :)

**Comment by alexjp** [comment]
> I dont know what it is currently most needed... in what areas can you help????? :)

Well ... i can try to test against weird configurations :)

I am also a programmer, but not very much into qml/qt programming. But would love to. ( my area is either web programming and windows wpf  )

**Comment by psifidotos** [comment]
> ell ... i can try to test against weird configurations :)
> this is more than a lot for now!!!! :) when I have a working "specific launcher per activity" I will ping you to test it!!!! :-)

thanks a lot!!!

**Comment by psifidotos** [comment]
@alexjp So.... :)
after a lot of hours... trying to debug javascript array behavior and stringlists ... I am exhausted!!! :)

but....

please @alexjp  test the scratch branch... it has support for multiple launchers between activities...

a launcher can be onAllActivities or just on specific ones... this can checked by the launchers right-click menu... it looks lovely in my system...

I hope this is ok for you!!! :)

I am sure it enhances a user workflow a lot!!!!

**Comment by alexjp** [comment]
Testing :)

So initial remarks :

The launcher on all activities/just specific activity seems great. Even worked as intented, when i added the same launcher in more than one activity and configured for specific activity.

Problems:
- The constant disappearing and appearing of the icons when switching activities are a pain :(
- I don't actually now where to put it now. Putting on nowdock.panel, it would be great if the panel were translucent

Can i make a new request ? Can you make nowdock behave as a panel ? ( meaning, it would show on all activities ?

Examples:

Usage with other plasmoids on normal panel, notice systray pushed up at top, not bottom. adding a spacer, makes it all break when dock adds more icons  -> http://imgur.com/a/dxgae

This looks bad -> http://imgur.com/a/uVlFH

Bottom line: I think if you could turn nowdock into a panel mode ( show all activities, keep below/autohide ) would be great. I am not being able to make it work very well with normal panel or nowdock.panel.

For now i will let it be as a "unity" clone :) with vertical panel on left.

Thanks !!

**Comment by alexjp** [comment]
How i have it setup for now -> http://imgur.com/a/50Y2U

The basic functionality is there. I guess its already enough for me to use it like that. So i have alot of activities where dolphin and konsole are "all activities launchers".

The others are relative to each activity.

**Comment by alexjp** [comment]
> I am sure it enhances a user workflow a lot!!!!

I have been waiting for this ( specific launchers for each activity ) since KDE 4.x !

I will even say that after this, people will start to understand why one should use activities and not virtual desktops to separate "activity contexts"

Thanks very much !

**Comment by psifidotos** [comment]
> The constant disappearing and appearing of the icons when switching activities are a pain :(

unfortunately this is based a lot on libtaskmanager... I could improve it I guess to only animate these that are different between the two activities (previous and current) and not all of them...
(open a different ticket if you want this)...

![nowdock-12](https://cloud.githubusercontent.com/assets/2216474/18234586/703120f2-7313-11e6-8080-8ebfd94a0ebe.png)

this is my Now Dock panel with breeze-dark theme...
The transparency for the panel can be supported through the theme...
You can download my script and its gui: [kdelook.org](https://dl.opendesktop.org/api/files/download/id/1473029935/plasma-transparent-panel.py.v0.2.2.tar.gz)
extract it and run with just a click **transparent.sh** this is its gui...

it opens a window and you can select your theme, which panel you want transparent and what shadows to show for your panels.
click at the bottom **"Generate Theme"** and this creates a new theme for you with just the specific panel transparent.
You can enable that theme as a regular theme for the systemsettings...
it will be named for example **Breeze Dark - West Transparrent** (this is the left panel transparent)

from Plasma 5.9 the previous steps may not be needed, it depends from the plasma developers...

The Now Dock Panel animates also many of its widgets if you want to... That creates a very consistant behavior with the plasmoid in it...

I hoped that helped... :)

Now Dock (Panel and Plasmoid) is a lot more mature that I had in my mind actually :)

**Comment by alexjp** [comment]
> unfortunately this is based a lot on libtaskmanager... I could improve it I guess to only animate these that are different between the two activities (previous and current) and not all of them...
> (open a different ticket if you want this)...

Not needed, I guess it would be really great if it was fixed upstream. I also have this problem on other issues on normal taskbar.

**Comment by alexjp** [comment]
About the transparent panel, yes, it really makes all the difference :)

But i actually only want to put nowdock as dock. Don't want to make a super busy dock.

This is comming along very nicelly !

**Comment by alexjp** [comment]
Noticed a problem with transparent panel: Although it makes the panel transparent, it still "blurs" the background.

**Comment by psifidotos** [comment]
@alexjp  blurring is based on KWin which blurs all transparent windows or none... To disable blurring you can just disable it from the Kwin effects, of course this means that it will not blur any transparency...But I dont think this is something to be solved upstream very very soon.... :)

**Comment by alexjp** [comment]
Hi,

Well, i think i will close this issue. It would be wondefull if it could work as a panel without those "background problems" but for now, i am pretty happy with the "unity layout" :)

Thanks a lot !!!

**Comment by psifidotos** [comment]
Ok!!! :)

@alexjp, you use blur in your windows and theme, right?

**Comment by alexjp** [comment]
Yes... I kinda wanna keep blur effect on the desktop. Just not on the panel.

I tried to set in kwin the panel as "screen", and it seemed to remove blur, but could not reproduce it after logging in and logging out.

Anyway, i think I were making requests and talking about issues that go beyond what I originally wanted. For my initial goal, it is acomplished :)

**Comment by psifidotos** [comment]
> I tried to set in kwin the panel as "screen", 

how did you do that?

dont worry about requests etc.... plasma has its limitations...

**Comment by alexjp** [comment]
tried to use kwin windows rules. I noticed when setting the panel to "screen" the blur disappeared !
